### DISCUSSION TITLE
Database session in router level ( using dependency injection )

### DISCUSSION THREAD

**Discussion by BH-ITAY** [fastapi/fastapi#5997] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@router.post("", status_code=status.HTTP_201_CREATED, response_model=ItemResponse)
async def create_item(
    payload: ItemSchema,
    db_session: AsyncSession = Depends(item_db.db.get_session),
    tenant_id: str = Depends(get_global_tenant_id),
):
    logging.debug(f"Handling request from tenant: {tenant_id}")
    item = await ItemController().create(db_session, payload.dict())
    return item

------------------------------ OR ------------------------------------------

@router.post("", status_code=status.HTTP_201_CREATED, response_model=ItemResponse)
async def create_item(
    payload: ItemSchema,
    tenant_id: str = Depends(get_global_tenant_id),
):
    logging.debug(f"Handling request from tenant: {tenant_id}")
    item = await ItemController().create(payload.dict())
    return item
```

### Description

This is more of a micro-service design challenge I came across and wanted to hear the thoughts of the crowd.

When using FastApi ```Depends``` injection, I get very good performance and I also love the idea that the session will be closed only after the response has been sent.

However, keeping the database session at the router level will be a bit challenging when building more complex micro-services ( with more than just basic CRUD operations, e.g. talk to an outside API and then create the object ).
The reason is that we get very high coupling between the router, the controller (business logic), and the model layer. Therefore testing will be coupled, any extension will be coupled, and generally extending the service will require more effort.

Do you guys still suggest keeping the db_session at the router level and just moving it around between the service layers ?

Thanks in advance,
IBH

### Operating System

Linux, Windows, macOS, Other

### FastAPI Version

0.89

### Python Version

3.11

**Comment by frankie567** [comment, marked as answer]
Yes, I would still recommend to have a session dependency. As you said, it clearly helps to have a single session opened when the request starts and closed when the response is sent.

Maybe what bugs you in your code is that you need to pass the session at each method call. We can make it slightly better by passing it to the initializer of `ItemController`.

```py
class ItemController:
    def __init__(self, session: AsyncSession):
        self.session = session

    async def create(self, payload) -> Item:
        item = Item(**payload)
        self.session.add(item)
        await self.session.commit()
```

The good thing with this approach is that you can even make `ItemController` injectable by a dependency (dependencies can be nested):

```py
async def get_item_controller(session: AsyncSession = Depends(get_session)) -> ItemController:
    return ItemController(session)


@router.post("", status_code=status.HTTP_201_CREATED, response_model=ItemResponse)
async def create_item(
    payload: ItemSchema,
    controller: ItemController = Depends(get_item_controller),
    tenant_id: str = Depends(get_global_tenant_id),
):
    logging.debug(f"Handling request from tenant: {tenant_id}")
    item = await controller.create(payload.dict())
    return item
```

The great advantage with this approach is when it comes to unit testing: with [`dependency_overrides`](https://fastapi.tiangolo.com/advanced/testing-dependencies/#use-the-appdependency_overrides-attribute) you can easily swap this controller with a fully mocked class.

`ItemController` is not tied to the router logic. If you need to use it in isolation, all you need is to instantiate a DB session:

```py
engine = create_async_engine(...)
async_session = async_sessionmaker(engine, expire_on_commit=False)
async with async_session() as session:
    controller = ItemController(session)
    item = await controller.get_by_id(123)
```

No FastAPI involved in the snippet above, only pure SQLAlchemy constructs (which could even be abstracted if you really need it).

**Reply by BH-ITAY** [reply]
Words to live by :)

Thanks for the deep look and the elaborated explanation.
